feat: Add Zoom connTest function

mario99logic opened the pull request:
- Implements OAuth and Server-to-Server authentication for Zoom API
- Adds connection tests to verify API access
- Includes token refresh logic for OAuth authentication

mario99logic committed:
add zoom test function

mario99logic committed:
fixes

tebeka commented on integrations/zoom/auth.go:
```diff
@@ -2,12 +2,123 @@ package zoom
+	data.Set("grant_type", "account_credentials")
+	data.Set("account_id", app.AccountID)
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodPost,
```
Are you sure this is how it works? Usually `url.Values` are used as URL query parameters.

```go
	query := url.Values{}
	query.Set("grant_type", "account_credentials")
	query.Set("account_id", "mcscrooge")
	u, _ := url.Parse("https://zoom.us/oauth/token")
	u.RawQuery = query.Encode()
        authURL := u.String()
```

daabr replied:
@tebeka Yes, afaik This Is The Way to construct a URL-encoded web form in a POST request: https://stackoverflow.com/questions/19253469/make-a-url-encoded-post-request-using-http-newrequest

tebeka commented on integrations/zoom/auth.go:
```diff
@@ -2,12 +2,123 @@ package zoom
+	req.SetBasicAuth(app.ClientID, app.ClientSecret)
+	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+
+	client := &http.Client{Timeout: 10 * time.Second}
```
Why 10 seconds? Also, the recommended way to set a timeout is to use `http.NewRequestWithContext`

```go
	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	req, err := http.NewRequestWithContext(ctx, http.MethodGet, "http://example.com", nil)
	if err != nil {
		return err
	}
	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return err
	}
	fmt.Println(resp.StatusCode)
	return nil

```

tebeka commented on integrations/zoom/vars.go:
```diff
@@ -8,3 +8,10 @@ type privateApp struct {
 	SecretToken  string `var:"private_secret_token,secret"`
 }
+
+type tokenResp struct {
```
Since you use it only once to parse JSON, use an anonymous struct in `serverToken`

daabr replied:
@mario99logic there are connection vars, so it makes sense to define the in this file,  but `tokenResp` is not a meaningful scope for these vars in this integration.

daabr replied:
Instead of defining this, you can reuse `common.OAuthData` - but you need to add the `json:"..."` annotations to `AccessToken`, `RefreshToken`, `Expiry` ("expires_in" in JSON).

See https://github.com/autokitteh/autokitteh/blob/main/integrations/microsoft/connection/vars.go for an example on how to do both `json` and `var` annotations.

No need to add JSON annotations to the other fields not mentioned out - you don't use them anyway.

tebeka commented on integrations/zoom/auth.go:
```diff
@@ -2,12 +2,123 @@ package zoom
+		Timeout: 10 * time.Second,
+	}
+
+	resp, err := client.Do(req)
```
You can create one client globally and then use it all over.
Also: Why use pointer and not value?

tebeka commented on integrations/zoom/auth.go:
```diff
@@ -2,12 +2,123 @@ package zoom
+func refreshTokenReq(ctx context.Context, refreshToken string, clientID string, clientSecret string, vs sdktypes.Vars) (string, error) {
+	data := url.Values{}
+	data.Set("grant_type", "refresh_token")
+	data.Set("refresh_token", refreshToken)
```
See comment about on `url.Values`

daabr commented on integrations/zoom/auth.go:
```diff
@@ -2,12 +2,123 @@ package zoom
+	"go.autokitteh.dev/autokitteh/sdk/sdktypes"
+)
+
+var (
```
Connection variables should be defined in the `vars.go` file

daabr commented on integrations/zoom/auth.go:
```diff
@@ -2,12 +2,123 @@ package zoom
+	data.Set("account_id", app.AccountID)
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodPost,
+		"https://zoom.us/oauth/token", strings.NewReader(data.Encode()))
```
No need to break this line. If you want, you can use a helper var for the body.

daabr commented on integrations/zoom/auth.go:
```diff
@@ -2,12 +2,123 @@ package zoom
+	return tokenData.AccessToken, nil
+}
+
+func makeAPICall(ctx context.Context, token string, url string) (*http.Response, error) {
```
1. This function is just a generic wrapper for an HTTP GET request, there's nothing specific for API calls, so the name is misleading - please rename to `httpGet`
2. There's nothing specifically related to authentication (other than using a bearer token), and it's not used anywhere in this file, so the location of this function is not right - please move to `checks.go`
3. With all the token types being used in this integration, it's worthwhile being explicit about the type/purpose/usage of the `token` argument to avoid confusion - please rename it to `bearerToken` even though it's longer (usually being verbose is bad, but in this case it's helpful)
4. This function isn't documented, see https://tip.golang.org/doc/comment#func

daabr commented on integrations/zoom/auth.go:
```diff
@@ -2,12 +2,123 @@ package zoom
+}
+
+// refreshTokenReq sends a request to Zoom's OAuth endpoint to refresh an expired access token.
+func refreshTokenReq(ctx context.Context, refreshToken string, clientID string, clientSecret string, vs sdktypes.Vars) (string, error) {
```
1. `Req` in the need is redundant
2. You're using a `privateApp` above, why not here too?

daabr commented on integrations/zoom/auth.go:
```diff
@@ -2,12 +2,123 @@ package zoom
+
+// refreshTokenReq sends a request to Zoom's OAuth endpoint to refresh an expired access token.
+func refreshTokenReq(ctx context.Context, refreshToken string, clientID string, clientSecret string, vs sdktypes.Vars) (string, error) {
+	data := url.Values{}
```
Notice that you're doing something very similar above - you can extract the common code into an `httpPost` helper function

daabr commented on integrations/zoom/auth.go:
```diff
@@ -2,12 +2,123 @@ package zoom
 // internal app details (based on: https://developers.zoom.us/docs/internal-apps/s2s-oauth/).
-func serverToken(_ context.Context, _ *privateApp) (*oauth2.Token, error) {
-	return nil, nil // TODO(INT-247): Implement.
+func serverToken(ctx context.Context, app *privateApp) (string, error) {
```
There's no reason to pass a reference to `privateApp` - it can't (i.e. it must not) be `nil`, and you're not modifying it

mario99logic replied:
I passed a reference to avoid modifying the original function, as it is used in the Save.go file. I will update it in the next push to demonstrate what I mean.

daabr commented on integrations/zoom/auth.go:
```diff
@@ -2,12 +2,123 @@ package zoom
+
+	var tokenData tokenResp
+
+	if err := json.NewDecoder(resp.Body).Decode(&tokenData); err != nil {
```
Merely declaring an empty var shouldn't be a separate block of code
```suggestion
	var tokenData tokenResp
	if err := json.NewDecoder(resp.Body).Decode(&tokenData); err != nil {
```

daabr commented on integrations/zoom/auth.go:
```diff
@@ -2,12 +2,123 @@ package zoom
+		return "", errors.New("failed to refresh token")
+	}
+
+	var tokenResp struct {
```
Same comment as the one about `tokenResp` in `vars.go`.

Regardless, there's no logic in defining this struct twice, in different files, both inside and outside functions that use it.

daabr commented on integrations/zoom/auth.go:
```diff
@@ -2,12 +2,123 @@ package zoom
+		return "", fmt.Errorf("failed to decode token response: %v", err)
+	}
+
+	if err := vs.Set(common.OAuthAccessTokenVar, tokenResp.AccessToken, true); err != nil {
```
Whenever you save a new access token, you also need to save its expiration timestamp

daabr commented on integrations/zoom/checks.go:
```diff
@@ -46,13 +49,96 @@ func test(v sdkservices.Vars) sdkintegrations.OptFn {
+	clientSecret := os.Getenv("ZOOM_CLIENT_SECRET")
+	url := "https://api.zoom.us/v2/users/me"
+
+	// First attempt with current token
```
Remember: if a comment is a full sentence, use punctuation, i.e. end with "."

daabr commented on integrations/zoom/checks.go:
```diff
@@ -46,13 +49,96 @@ func test(v sdkservices.Vars) sdkintegrations.OptFn {
+		return sdktypes.NewStatus(sdktypes.StatusCodeOK, "OAuth connection successful"), nil
+	}
+
+	// Check if token expired
```
Discussed offline: compare the OAuth expiration to the current timestamp, instead of making the same API call twice

daabr commented on integrations/zoom/checks.go:
```diff
@@ -46,13 +49,96 @@ func test(v sdkservices.Vars) sdkintegrations.OptFn {
+	return sdktypes.NewStatus(sdktypes.StatusCodeError, "Failed to connect to Zoom API"), nil
+}
+
+// ServerToServerTest handles the Server-to-Server authentication test for Zoom
```
Don't use the word "handle" unless you really have to, 99% of the time there are more accurate and descriptive verbs.

daabr commented on integrations/zoom/checks.go:
```diff
@@ -46,13 +49,96 @@ func test(v sdkservices.Vars) sdkintegrations.OptFn {
+	clientID := vs.GetValue(clientID)
+	secret := vs.GetValue(clientSecret)
+
+	app := privateApp{
```
Discussed offline: use `vs.Decode`

mario99logic replied:
changed it just kept the os.Getenv

daabr commented on integrations/zoom/checks.go:
```diff
@@ -46,13 +49,96 @@ func test(v sdkservices.Vars) sdkintegrations.OptFn {
+	}
+
+	url := "https://api.zoom.us/v2/users"
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
```
Why not use `makeAPICall` (or `httpGet` according to my comment about this function name)

mario99logic committed:
fixes

tebeka commented:
BTW: The PR title is not relevant.

mario99logic committed:
fixes

mario99logic committed:
typo fix

mario99logic commented:
notes for the changes:

- Implemented the fix for the Decode function that we discussed offline.
- In `oauthTest`, I use Decode, but I still retrieve `clientID` and `clientSecret` from environment variables since we don’t receive them from the client.
- In `refreshToken`, I introduced a new struct instead of using `common.OAuthData` because the expiry field in `common.OAuthData` is a string, whereas the JSON response provides it as an integer.
- Although the JSON response currently returns the same `expires_in` value each time, I prefer to retrieve it from the response to accommodate any potential future changes.

daabr commented on sdk/sdktypes/vars.go:
```diff
@@ -113,12 +113,18 @@ func (vs Vars) Decode(out any) {
 		fv := v.Field(i)
 		ft := t.Field(i)
 
+		tag := strings.Split(ft.Tag.Get("var"), ",")
```
FYI - I already copied this code (and added unit tests) in https://github.com/autokitteh/autokitteh/pull/1164/files#diff-4f8b0dd502d810a388ec87745edc8ac98e0d5b734258a4420dbc27bdecf24fd9

You can remove the changes in this file in this PR, and rebase with the main branch.

daabr commented on integrations/zoom/checks.go:
```diff
@@ -46,13 +52,116 @@ func test(v sdkservices.Vars) sdkintegrations.OptFn {
+
+// httpPost sends an HTTP POST request with a URL-encoded body and basic authentication.
+func httpPost(ctx context.Context, data url.Values, clientID, clientSecret string) (*http.Response, error) {
+	timeoutCtx, cancel := context.WithTimeout(ctx, 20*time.Second)
```
Nit: you're not using `ctx` anywhere else in this function, so you can just overwrite `ctx`: this will replace the input, but only inside this function, which is what you want.

Anyway, try to model this function more or less like a simplified, Zoom-specific copy of `httpRquest` from https://github.com/autokitteh/autokitteh/blob/main/integrations/common/http.go#L98

daabr commented on integrations/zoom/checks.go:
```diff
@@ -46,13 +52,116 @@ func test(v sdkservices.Vars) sdkintegrations.OptFn {
+	data := common.OAuthData{}
+	vs.Decode(&data)
+	clientID := os.Getenv("ZOOM_CLIENT_ID")
+	clientSecret := os.Getenv("ZOOM_CLIENT_SECRET")
```
Notice that this means you don't support private OAuth here yet.
Add a TODO comment with a ticket reference to fix that later.

daabr commented on integrations/zoom/checks.go:
```diff
@@ -46,13 +52,116 @@ func test(v sdkservices.Vars) sdkintegrations.OptFn {
+	clientSecret := os.Getenv("ZOOM_CLIENT_SECRET")
+	url := "https://api.zoom.us/v2/users/me"
+
+	// Check if the token is expired and refresh it.
```
Instead of this logic, use the new function `FreshOAuthToken` from https://github.com/autokitteh/autokitteh/blob/main/integrations/common/vars.go#L84

daabr commented on integrations/zoom/checks.go:
```diff
@@ -46,13 +52,116 @@ func test(v sdkservices.Vars) sdkintegrations.OptFn {
+}
+
+// httpGet sends an HTTP GET request to the specified URL with a Bearer token for authentication.
+func httpGet(ctx context.Context, bearerToken string, url string) (*http.Response, error) {
```
Please replace `httpGet` here with `common.HTTPGet` (but not `httpPost` below, because it doesn't support basic auth). See also INT-337 (for a future PR).

daabr commented on integrations/zoom/auth.go:
```diff
@@ -2,12 +2,75 @@ package zoom
+}
+
+// refreshToken sends a request to Zoom's OAuth endpoint to refresh an expired access token.
+func refreshToken(ctx context.Context, refreshT string, clientID string, clientSecret string, vs sdktypes.Vars) (string, error) {
```
Don't refresh directly, instead use `common.FreshOAuthData` from https://github.com/autokitteh/autokitteh/blob/main/integrations/common/vars.go#L84

mario99logic committed:
fixes

mario99logic committed:
typo fix

mario99logic committed:
lint fix

daabr commented on integrations/zoom/checks.go:
```diff
@@ -46,13 +53,74 @@ func test(v sdkservices.Vars) sdkintegrations.OptFn {
+func oauthTest(ctx context.Context, v sdkservices.Vars, o sdkservices.OAuth, vs sdktypes.Vars) (sdktypes.Status, error) {
+	// TODO(INT-338): Support private OAuth.
+	// Check if the token is expired and refresh it.
+	desc := common.Descriptor("zoom", "", "")
```
No need to redefine `desc` here, this file already has access to the `desc` global var in `integration.go`

daabr commented on integrations/zoom/checks.go:
```diff
@@ -46,13 +53,74 @@ func test(v sdkservices.Vars) sdkintegrations.OptFn {
+
+// ServerToServerTest validates the Server-to-Server authentication test for Zoom.
+// (Based on: https://developers.zoom.us/docs/internal-apps/s2s-oauth/).
+func ServerToServerTest(ctx context.Context, vs sdktypes.Vars) (sdktypes.Status, error) {
```
Why is this function exported? Is it used in any other package?

mario99logic replied:
no need to export it, changed it.

daabr reviewed (approved):
Approved with 2 easy-to-fix comments

mario99logic committed:
fixes
